### ISSUE TITLE
InternVL-Chat-V1.2-Plus  能提供量化模型吗

### ISSUE THREAD

**Issue by sunjunlishi** [OpenGVLab/InternVL#82]
InternVL-Chat-V1.2-Plus  根据您们提供的demo，这是我测试过所有的开源的（甚至仅提供demo的）最好的模型！！！
能提供量化的模型吗？因为这个模型文件共计80G，普通卡跑不起来；

如果量化为4bit，或者2bit的gguf格式的文件，可以普通显卡就可以跑。
而且越大大模型量化，精度损失越小。
希望能转化下，更加以飨读者。

**Comment by sunjunlishi** [comment]
![1712047441545](https://github.com/OpenGVLab/InternVL/assets/11845664/86ecd4fd-7409-49a6-8e15-5107b005c411)
大模型的量化模型用来训练更有性价比

**Comment by czczup** [comment]
感谢您的建议。我最近会转换一下量化模型并上传到hugging face。

**Comment by sunjunlishi** [comment]
还有奥，qwen-vl-plus分辨率比较高，咱们当前是448，如果能再增加下分辨率比如996像素，应该会更加强劲。
还有咱们这应该没有专门针对测试数据，因为我实测身份证和营业执照等关键信息提取，都非常好，star应该上万，其他人没有发现此处是宝。

**Comment by czczup** [comment]
> 咱们的(原来是v1-2plus已基本够用,版本3输出位置是按照1000*1000)，很强。 还有咱们这应该没有专门针对测试数据，因为我实测身份证和营业执照等关键信息提取，都非常好，star应该上万，其他人没有发现此处是宝。

感谢反馈！目前已经训好了支持到2K-4K分辨率的模型，刷点性能和实际体验都有大幅提升，预计下周会放出来，可能这两天会更新一个预览的demo。

**Comment by sunjunlishi** [comment]
@czczup 哇塞效果这么好。普通消费者只能下载 10G左右的模型；

**Comment by czczup** [comment]
> @czczup 哇塞效果这么好。普通消费者只能下载 10G左右的模型；

有试试V1.5的demo吗

**Comment by sunjunlishi** [comment]
@czczup 对的，印章，手写，关键信息提取，都非常棒。我只能先跑下特征提取那部分了。量化的需要合作？

**Comment by sunjunlishi** [comment]
https://huggingface.co/failspy/InternVL-Chat-V1-5-4bit 
这里提供了一个量化的模型，就是不知道怎么调用

**Comment by sunjunlishi** [comment]
@czczup  老师，给转换一下呗，需要测试数据进行压缩是吗？

**Comment by sunjunlishi** [comment]
import json
import os

import torch
import torchvision.transforms as T
from PIL import Image
from torchvision.transforms.functional import InterpolationMode
from tqdm import tqdm
from transformers import AutoModel, AutoTokenizer

from torchvision.transforms.functional import InterpolationMode


IMAGENET_MEAN = (0.485, 0.456, 0.406)
IMAGENET_STD = (0.229, 0.224, 0.225)


def build_transform(input_size):
    MEAN, STD = IMAGENET_MEAN, IMAGENET_STD
    transform = T.Compose([
        T.Lambda(lambda img: img.convert('RGB') if img.mode != 'RGB' else img),
        T.Resize((input_size, input_size), interpolation=InterpolationMode.BICUBIC),
        T.ToTensor(),
        T.Normalize(mean=MEAN, std=STD)
    ])
    return transform


def find_closest_aspect_ratio(aspect_ratio, target_ratios, width, height, image_size):
    best_ratio_diff = float('inf')
    best_ratio = (1, 1)
    area = width * height
    for ratio in target_ratios:
        target_aspect_ratio = ratio[0] / ratio[1]
        ratio_diff = abs(aspect_ratio - target_aspect_ratio)
        if ratio_diff < best_ratio_diff:
            best_ratio_diff = ratio_diff
            best_ratio = ratio
        elif ratio_diff == best_ratio_diff:
            if area > 0.5 * image_size * image_size * ratio[0] * ratio[1]:
                best_ratio = ratio
    return best_ratio


def dynamic_preprocess(image, min_num=1, max_num=6, image_size=448, use_thumbnail=False):
    orig_width, orig_height = image.size
    aspect_ratio = orig_width / orig_height

    # calculate the existing image aspect ratio
    target_ratios = set(
        (i, j) for n in range(min_num, max_num + 1) for i in range(1, n + 1) for j in range(1, n + 1) if
        i * j <= max_num and i * j >= min_num)
    target_ratios = sorted(target_ratios, key=lambda x: x[0] * x[1])

    # find the closest aspect ratio to the target
    target_aspect_ratio = find_closest_aspect_ratio(
        aspect_ratio, target_ratios, orig_width, orig_height, image_size)

    # calculate the target width and height
    target_width = image_size * target_aspect_ratio[0]
    target_height = image_size * target_aspect_ratio[1]
    blocks = target_aspect_ratio[0] * target_aspect_ratio[1]

    # resize the image
    resized_img = image.resize((target_width, target_height))
    processed_images = []
    for i in range(blocks):
        box = (
            (i % (target_width // image_size)) * image_size,
            (i // (target_width // image_size)) * image_size,
            ((i % (target_width // image_size)) + 1) * image_size,
            ((i // (target_width // image_size)) + 1) * image_size
        )
        # split the image
        split_img = resized_img.crop(box)
        processed_images.append(split_img)
    assert len(processed_images) == blocks
    if use_thumbnail and len(processed_images) != 1:
        thumbnail_img = image.resize((image_size, image_size))
        processed_images.append(thumbnail_img)
    return processed_images


def load_image(image_file, input_size=448, max_num=6):
    image = Image.open(image_file).convert('RGB')
    transform = build_transform(input_size=input_size)
    images = dynamic_preprocess(image, image_size=input_size, use_thumbnail=True, max_num=max_num)
    pixel_values = [transform(image) for image in images]
    pixel_values = torch.stack(pixel_values)
    return pixel_values



path = "/workspace/workspace/Llama/InternVL-Chat-V1-5-4bit"
# If you have an 80G A100 GPU, you can put the entire model on a single GPU.
model = AutoModel.from_pretrained(
    path,
    load_in_4bit=True,
     torch_dtype=torch.bfloat16,
    low_cpu_mem_usage=True,
    trust_remote_code=True,
    device_map='auto').eval()
# Otherwise, you need to set device_map='auto' to use multiple GPUs for inference.
# model = AutoModel.from_pretrained(
#     path,
#     torch_dtype=torch.bfloat16,
#     low_cpu_mem_usage=True,
#     trust_remote_code=True,
#     device_map='auto').eval()

from transformers import AutoModelForCausalLM, BitsAndBytesConfig, AutoTokenizer
 
# quantization configuration for NF4 (4 bits)
quantization_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type='nf4',
    bnb_4bit_compute_dtype=torch.bfloat16
)


#model1 = AutoModelForCausalLM.from_pretrained(path, load_in_4bit=True, device_map="auto")

tokenizer = AutoTokenizer.from_pretrained(path, 
    load_in_4bit=True,
    bnb_4bit_quant_type='nf4',
    bnb_4bit_compute_dtype=torch.bfloat16,
    trust_remote_code=True)
# set the max number of tiles in `max_num`
pixel_values = load_image('./test1.jpg', max_num=6).to(torch.bfloat16).cuda()

generation_config = dict(
    num_beams=1,
    max_new_tokens=512,
    do_sample=False,
)

# single-round single-image conversation
question = "分析这张图片内容"
response = model.chat(tokenizer, pixel_values, question, generation_config)

#response1 = model1.chat(tokenizer, pixel_values, question, generation_config)


print(question, response)
print("response:",response)


可以加载推理但是空

**Comment by sunjunlishi** [comment]
https://github.com/OpenGVLab/InternVL/issues/102 继续跟踪调用代码。

**Comment by sunjunlishi** [comment]
@czczup 老师呀，那个int8量化的还是不中呀，我们两张24G的卡。单张显存不够，两张说，1张加载，一张跑，张量不匹配的错误。

**Comment by leeaction** [comment]
请问4bit量化后的模型InternVL-Chat-V1-5-4bit，大约最少需要多少显存？

**Comment by Single430** [comment]
> @czczup 老师呀，那个int8量化的还是不中呀，我们两张24G的卡。单张显存不够，两张说，1张加载，一张跑，张量不匹配的错误。

应该要设置device_map=“auto”吧，4090x2测试可用

**Comment by Single430** [comment]
@czczup 你好，internVL-V1.5目前也是用过最好的图像理解模型，但是开源的资料比较少，在llama.cpp中也发现还没有支持转gguf，而且也未看到stream_chat的相应代码，不知道你们是否有做相应工作？

**Comment by sunjunlishi** [comment]
from lmdeploy import pipeline
from lmdeploy.vl import load_image
pipe = pipeline('../Mini-InternVL-Chat-2B-V1-5')
image = load_image('fapiao.jpg')

import time
 
start = time.time()
response = pipe(('请提取购买方的名称、纳税人识别号、地址、电话、开户行及账号；以及销售方的名称、纳税人识别号、地址、电话、开户行及账号信息', image))
end = time.time()
print(f"耗时: {end - start}秒")

print(response)  验证可以

**Comment by sunjunlishi** [comment]
总结 ../Mini-InternVL-Chat-2B-V1-5         4.5G 显存占用13G，推理发票2.8s；推理结果稳定，但是字符有错误。是同等里面我感觉还是最好的。
优化：结合ocr识别作为其上下文，结果非常完美。
再次总结：ocr识别作为上下文+Mini-InternVL-Chat-2B-V1-5 效果比大的多模态效果还好(前提你的ocr识别要好)；
效果也比ocr+大语言模型效果好。
整体来说，我比较喜欢这个小的多模态。

**Comment by sunjunlishi** [comment]
ocr训练可控。加上小的多模态，效果还是非常震感，非常感谢作者老师们的工作。只能感谢。
